hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -0,0 +1,108 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+
+import scala.collection.mutable
+
+object ForwardLets {
+  type Strategy = Int
+  private val ELIMINATE: Strategy = 0
+  private val FORWARD: Strategy = 1
+  private val PROCEED: Strategy = 2
+
+  private def rewriteIR(ir0: IR): IR = {
+    val UsesAndDefs(uses, defs) = ComputeUsesAndDefs(ir0)
+    val nestingDepth = NestingDepth(ir0)
+
+    val m = Memo.empty[IR]
+
+    var needsRewrite = true
+
+    def strategy(x: RefEquality[IR], refs: mutable.Set[RefEquality[Ref]]): Strategy = {
+      val size = refs.size
+
+      if (size == 0)
+        ELIMINATE
+      else if (size == 1) {
+        val depth = nestingDepth.lookup(x)
+        if (nestingDepth.lookup(refs.head) == nestingDepth.lookup(x))
+          FORWARD
+        else
+          PROCEED
+      } else PROCEED
+    }
+
+    def rewrite(ir: IR): IR = {
```

Comment:
I'm slightly confused by this pattern.  I would have expected a top-down rewrite that carries `Env[IR]` of replacements.  When you hit a let, if it is empty you immediately delete it, if has one occurrence, and if more than one then do nothing.  I think this will clean things up greatly unless I'm missing something.  I think this will clean things up (get rid of the enumerated strategy, etc.)

Reply:
If I use an `Env[IR]`, then I need to handle all of the name-binding-and-collision logic in this method. That is going to be much more complex, no?